service/src/event/validator.rs

```diff
@@ -0,0 +1,269 @@
+use std::collections::HashMap;
+
+use ceramic_core::StreamId;
+use ceramic_event::unvalidated::{self, init, signed};
+use ceramic_store::{CeramicOneEvent, EventInsertable, SqlitePool};
+use ceramic_validation::{cacao_verifier::Verifier, event_verifier::Verifier as _};
+use cid::Cid;
+use ipld_core::ipld::Ipld;
+use tokio::try_join;
+
+use super::{service::InvalidItem, DeliverableRequirement};
+
+use crate::Result;
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub enum HistoryRequirement {
+    Complete,
+    InitOnly,
+    None,
+}
+
+impl From<DeliverableRequirement> for HistoryRequirement {
+    fn from(value: DeliverableRequirement) -> Self {
+        match value {
+            DeliverableRequirement::Immediate => HistoryRequirement::Complete,
+            DeliverableRequirement::Asap => HistoryRequirement::InitOnly,
+            DeliverableRequirement::Lazy => HistoryRequirement::None,
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct EventValidator<'a> {
+    pool: &'a SqlitePool,
+    history_req: HistoryRequirement,
+}
+
+impl<'a> EventValidator<'a> {
+    pub fn new(pool: &'a SqlitePool, history_req: HistoryRequirement) -> Self {
+        Self { pool, history_req }
+    }
+
+    pub(crate) async fn validate_events(
+        &self,
+        parsed_events: Vec<EventInsertable>,
+        invalid_events: &mut Vec<InvalidItem>,
+    ) -> Result<(Vec<EventInsertable>, Vec<EventInsertable>)> {
+        // first collect all the init events so we can refer to them when grouping by type
+        // in order to avoid querying the database or rejecting the events incorrectly.
+        // we will then add any missing init events from the DB for the next phase.
+        // guess at a capacity and just assume half are init events so we only need to re-allocate once at most
+        // TODO: we should validate these init events before we trust them for others.. could
```

Comment:
what's the plan for addressing this TODO?  Do we need a ticket for this?

Reply:
maybe it should be a ticket.. I was thinking to just do things in two passes initially: split init events out and check their signatures before using them in the next phase.. it's a bit of a bummer but isn't hard, I just didn't want to forget so I wrote this TODO.

Reply:
added in this PR